Add group ACL rule for membership in another group

jboning opened the pull request:
This will make it possible to configure different API key requirements for different alliances.

jboning committed:
add ACL model for recursive group membership

jboning committed:
add group membership ACL rule to UI

jboning committed:
ACLGroup fixes

 - respect grant/deny and inverse
 - __unicode__ is how we're stringifying now

jboning committed:
group acl improvements

 - check for cycles before saving (not perfect because no database
   locking)
 - not using a global for cycle detection
 - naming tweaks

jboning committed:
what did I have contextmanager for

jboning committed:
clean up extra newline

jboning committed:
Be a little bit more flexible about passed-in context

Acidity commented on brave/core/group/controller.py:
```diff
@@ -129,20 +129,33 @@ def listify(rule, field):
             self.group.rules = rule_objects
+
+        try:
+            self.group.recursion_check()
```
I believe this should be self.group.cycle_check()

Acidity commented:
When attempting to make a cyclic reference, the ACL check from the server just displays [object Object] rather than an error code.

Acidity commented:
Another thought, since this only lets you specify one group in the ACL, if you want to do a DENY OR or a GRANT AND on multiple groups you have to make a new group just for that, which isn't great. Not sure how easy it would be to make the ACL support multiple groups though.

Acidity commented:
Also: deleting a group that another group refers to without deleting the ACL rule first breaks pretty much everything.

jboning committed:
Merge branch 'develop' into feature/acl-group

jboning committed:
Fixes

 - fix call that I missed when renaming
 - check for references in ACLs before deleting
 - tweak error handling a bit

jboning committed:
guarantee (modulo lol mongo concurrency) that we won't have dangling group references

jboning committed:
guarantee (modulo lol mongo concurrency) no cyclic group references

jboning committed:
ACLGroupReference tests

jboning commented:
Note to self, don't merge yet, this can still result in dangling references if you change the ID of a referenced group.

jboning committed:
update references on rename
